### ISSUE TITLE
deadpool-postgres 0.9 fails with "No runtime specified"

### ISSUE THREAD

**Issue by godofdream** [bikeshedder/deadpool#101]
0.9 has braking changes. I get "No runtime specified" while using tokio 1.
```
#deadpool-postgres = {version="0.7"} # old
deadpool-postgres = {version="0.9", features=["config","rt_tokio_1"]}
```

Any Idea?

**Comment by bikeshedder** [comment]
In `deadpool 0.9` I introduced the concept of a `Runtime` enum that can be used to select which sleep method should be used:
https://docs.rs/deadpool/0.8.0/deadpool/enum.Runtime.html

`deadpool_postgres` re-exports the `Runtime` enum so you can use it like that:
https://docs.rs/deadpool-postgres/0.9.0/deadpool_postgres/enum.Runtime.html

```rust
use std::time::Duration;
use deadpool_postgres::{Config, Manager, ManagerConfig, Pool, PoolConfig, RecyclingMethod, Runtime};
use tokio_postgres::NoTls;

#[tokio::main]
async fn main() {
    let mut cfg = Config::new();
    cfg.dbname = Some("deadpool".to_string());
    cfg.manager = Some(ManagerConfig { recycling_method: RecyclingMethod::Fast });
    let mut pool_cfg = PoolConfig::default();
    pool_cfg.runtime = Runtime::Tokio1;
    pool_cfg.timeouts.create = Duration::from_millis(100);
    pool_cfg.timeouts.wait = Duration::from_millis(1000);
    pool_cfg.timeouts.recycle = Duration::from_millis(100);
    cfg.pool = pool_config;
    for i in 1..10 {
        let mut client = pool.get().await.unwrap();
        let stmt = client.prepare_cached("SELECT 1 + $1").await.unwrap();
        let rows = client.query(&stmt, &[&i]).await.unwrap();
        let value: i32 = rows[0].get(0);
        assert_eq!(value, i + 1);
    }
}
```

That way it became fully compatible with `async-std`. See #68

If there is a way to "auto-detect" the current runtime at little cost I'd like to add a `Runtime::Auto` variant. I did not dig into this and felt like making it explicit was the best choice in the end anyways. The `Runtime` only needs to be specified if you're using timeouts.

btw. I excluded the runtime config from the deserialiation as I felt this should be a configuration defined at compile time rather than runtime. So specifying it like this does **NOT** work: `PG__POOL__RUNTIME=Tokio1`

**Comment by godofdream** [comment]
I assumed by setting the feature, the runtime would be set.
Many other crates switch the runtime using features.

Do you need the Runtime be selectable during runtime?
Or do you need multiple runtimes simultaneously?

**Comment by bikeshedder** [comment]
Mutually exclusive features in libraries is something that should usually be avoided:

- https://github.com/rust-lang/cargo/blob/44596fa2b18db6aa0c4820a5a89c31416aa2ad7b/src/doc/src/reference/features.md#mutually-exclusive-features
- https://github.com/rust-lang/cargo/issues/2980

When first looking into this I almost went with something like [agnostik](https://crates.io/crates/agnostik), but after talking to other Rust developers I got convinced that this is an anti-pattern and so I opted against it.

Maybe I should make the `Runtime` a required parameter for the `Config::create_pool` function in order to enforce a runtime selection by the user? The current API and README is far from optimal, I agree.

**Comment by bikeshedder** [comment]
fyi. There is nothing wrong with having multiple runtimes at the same time. It is actually possible to run a `tokio` and `async-std` executor in parallel. It's even possible to have two versions of `tokio` running at the same time.

Just imagine you want to use a `async-std`-only library your mostly `tokio` driven code. You could wrap that library in such a way that it has it's own `async-std` executor and communication is done via channels, semaphores and other synchronization primitives which are executor agnostic.

This is for sure no regular use case, but with a mutually exclusive runtime features this would be impossible to do.

**Comment by bikeshedder** [comment]
I also had the idea of making `Runtime::Tokio1` the default when `rt_tokio_1` is enabled. This seamed to be the best option for not breaking existing code wanting to update to `deadpool 0.8`, but I soon realized that this might cause other problems in the long run:

Just assume you're writing an application that uses `deadpool` without specifying a runtime and another crate that depends on `deadpool` with `rt_tokio_1`. Now you application is able to use tokio timeouts just fine since the feature got enabled indirectly. Let's asume that other crate gets updated and no longer depends on `rt_tokio_1`. Your code now fails at runtime, not compile time.

With the `Runtime` being explicit you would get a compile time error in this case as `Runtime::Tokio1` does not exist. Looking at the docs you get to know immediately that this feature is enabled using the `rt_tokio_1` feature.

**Comment by bikeshedder** [comment]
Thinking about this I guess it'd be best to change the interface of `deadpool` so that `Pool::from_config` returns a `Result<Pool, ConfigError>` which can check that timeouts can only be specified when also selecting a `Runtime`.

This is definitely a `deadpool 1.0` change and I've created a new tracking issue for that: #107 

This doesn't solve the case when someone creates a `Pool` without a timeout and later calls `Poll::get_timeout`, but then it should be pretty self explaining why it returns a `No runtime specified` error.

I will also add this to the documentation of `Pool::get_timeout`. When #107 is implemented `Pool::get` will be guaranteed to never return a `PoolError::NoRuntime`.

**Comment by bikeshedder** [comment]
This will be fixed by #110. The `PoolBuilder` will perform this validation and fail on pool creation if specifying timeouts without a runtime.
